File: src/lib/signals/strategies/momentum-ema.ts
import { adx, atr, closes, ema } from "../indicators";
import type { Strategy, StrategyContext, StrategySignal } from "../types";

const DEFAULTS = {
  fast: 9,
  slow: 21,
  adxPeriod: 14,
  adxThreshold: 20,
  atrPeriod: 14,
  atrStopMult: 2.0,
  rewardMult: 2.5, // target distance = rewardMult * stop distance
};

/**
 * EMA fast/slow crossover with ADX trend-strength filter.
 * Enters long when fast > slow AND ADX ≥ threshold AND fast is rising.
 * Enters short on the mirror condition. Exits to flat on crossover
 * reversal or ADX collapse below threshold.
 *
 * Stops are ATR-based; targets use a configurable R:R ratio.
 */
export function momentumEMA(overrides: Partial<typeof DEFAULTS> = {}): Strategy {
  const params = { ...DEFAULTS, ...overrides };
  const warmup = Math.max(params.slow + 2, params.adxPeriod * 2 + 1, params.atrPeriod + 2);

  return {
    id: "momentum-ema",
    params,
    warmupBars: warmup,
    onCandle(ctx: StrategyContext): StrategySignal | null {
      const candles = ctx.candles.toArray();
      if (candles.length < warmup) return null;

      const c = closes(candles);
      const { fast, slow, adxPeriod, adxThreshold, atrPeriod, atrStopMult, rewardMult } = params;

      const emaFast = ctx.indicators.get(`ema:close:${fast}`, () => ema(c, fast));
      const emaSlow = ctx.indicators.get(`ema:close:${slow}`, () => ema(c, slow));
      const adxOut = ctx.indicators.get(`adx:${adxPeriod}`, () => adx(candles, adxPeriod));
      const atrOut = ctx.indicators.get(`atr:${atrPeriod}`, () => atr(candles, atrPeriod));

      const i = candles.length - 1;
      const px = c[i];
      const f = emaFast[i];
      const s = emaSlow[i];
      const fPrev = emaFast[i - 1];
      const a = adxOut.adx[i];
      const t = atrOut[i];

      if (![f, s, fPrev, a, t].every(Number.isFinite)) return null;

      const diff = f - s;
      const slope = f - fPrev;
      const trendOn = a >= adxThreshold;

      const features: Record<string, number> = {
        emaFast: f,
        emaSlow: s,
        emaDiffPct: (diff / px) * 100,
        adx: a,
        plusDi: adxOut.plusDi[i],
        minusDi: adxOut.minusDi[i],
        atr: t,
        slope,
      };

      if (!trendOn) {
        return mkSignal("flat", 0, clamp((adxThreshold - a) / adxThreshold, 0, 1), ctx.now, features);
      }
      // Long regime
      if (diff > 0 && slope > 0) {
        const stop = px - atrStopMult * t;
        const target = px + atrStopMult * t * rewardMult;
        const score = tanh(diff / px * 50);
        return mkSignal("long", score, clamp(a / 50, 0, 1), ctx.now, features, { price: px, stop, target });
      }
      // Short regime
      if (diff < 0 && slope < 0) {
        const stop = px + atrStopMult * t;
        const target = px - atrStopMult * t * rewardMult;
        const score = tanh(diff / px * 50);
        return mkSignal("short", score, clamp(a / 50, 0, 1), ctx.now, features, { price: px, stop, target });
      }
      // Crossover but weak slope → flat
      return mkSignal("flat", 0, 0.2, ctx.now, features);
    },
  };
}

function mkSignal(
  side: StrategySignal["side"],
  score: number,
  confidence: number,
  ts: number,
  features: Record<string, number>,
  entryHint?: StrategySignal["entryHint"],
): StrategySignal {
  return { strategyId: "momentum-ema", ts, side, score, confidence, features, entryHint };
}

function clamp(x: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, x));
}

function tanh(x: number): number {
  if (x > 20) return 1;
  if (x < -20) return -1;
  const e2 = Math.exp(2 * x);
  return (e2 - 1) / (e2 + 1);
}

File: python/backtester/strategies/momentum_ema.py
"""Mirror of src/lib/signals/strategies/momentum-ema.ts."""
from dataclasses import dataclass, field
from typing import Any

import numpy as np

from ..indicators import adx, atr, ema


@dataclass
class MomentumEMA:
    fast: int = 9
    slow: int = 21
    adx_period: int = 14
    adx_threshold: float = 20.0
    atr_period: int = 14
    atr_stop_mult: float = 2.0
    reward_mult: float = 2.5
    params: dict[str, Any] = field(default_factory=dict)

    id: str = "momentum-ema"

    @property
    def warmup_bars(self) -> int:
        return max(self.slow + 2, self.adx_period * 2 + 1, self.atr_period + 2)

    def evaluate(self, h, l, c):
        n = len(c)
        ef = ema(c, self.fast)
        es = ema(c, self.slow)
        adx_out = adx(h, l, c, self.adx_period)
        a = atr(h, l, c, self.atr_period)

        side = np.full(n, "null", dtype=object)
        score = np.full(n, np.nan)
        confidence = np.full(n, np.nan)
        stops = np.full(n, np.nan)
        targets = np.full(n, np.nan)

        for i in range(1, n):
            f, s, fp, ax, t = ef[i], es[i], ef[i - 1], adx_out["adx"][i], a[i]
            if any(np.isnan(v) for v in (f, s, fp, ax, t)):
                continue
            diff = f - s
            slope = f - fp
            trend_on = ax >= self.adx_threshold
            if not trend_on:
                side[i] = "flat"
                score[i] = 0.0
                confidence[i] = max(0.0, min(1.0, (self.adx_threshold - ax) / self.adx_threshold))
                continue
            if diff > 0 and slope > 0:
                side[i] = "long"
                stops[i] = c[i] - self.atr_stop_mult * t
                targets[i] = c[i] + self.atr_stop_mult * t * self.reward_mult
                score[i] = np.tanh(diff / c[i] * 50)
                confidence[i] = max(0.0, min(1.0, ax / 50))
            elif diff < 0 and slope < 0:
                side[i] = "short"
                stops[i] = c[i] + self.atr_stop_mult * t
                targets[i] = c[i] - self.atr_stop_mult * t * self.reward_mult
                score[i] = np.tanh(diff / c[i] * 50)
                confidence[i] = max(0.0, min(1.0, ax / 50))
            else:
                side[i] = "flat"
                score[i] = 0.0
                confidence[i] = 0.2

        return {
            "side": side,
            "score": score,
            "confidence": confidence,
            "ema_fast": ef,
            "ema_slow": es,
            "adx": adx_out["adx"],
            "atr": a,
            "stop": stops,
            "target": targets,
        }
